### PR TITLE
hack/ci/windows.ps1: build busybox from local Dockerfile

### PR DESCRIPTION
**Pull request by thaJeztah**
this vendors the Dockerfile from http://github.com/jhowardmsft/busybox, and updates it to take build-time arguments

### CODE REVIEW

**Comment by thaJeztah** [comment]
@StefanScherer @andrewhsu ptal

File: hack/ci/windows.ps1 [hunk 1]
```diff
@@ -774,7 +774,7 @@ Try {
             #if ($bbCount -eq 0) {
                 Write-Host -ForegroundColor Green "INFO: Building busybox"
                 $ErrorActionPreference = "SilentlyContinue"
-                $(& "$env:TEMP\binary\docker-$COMMITHASH" "-H=$($DASHH_CUT)" build -t busybox https://raw.githubusercontent.com/jhowardmsft/busybox/v1.1/Dockerfile | Out-Host)
+                $(& "$env:TEMP\binary\docker-$COMMITHASH" "-H=$($DASHH_CUT)" build --build-arg WINDOWS_BASE_IMAGE --build-arg WINDOWS_BASE_IMAGE_TAG -t busybox "$env:SOURCES_DRIVE`:\$env:SOURCES_SUBDIR\src\github.com\docker\docker\contrib\busybox\Dockerfile" | Out-Host)
```
Comment (StefanScherer) [inline hack/ci/windows.ps1]: The build context is the directory, the `\Dockerfile` must be removed here.

**Review by StefanScherer — CHANGES_REQUESTED** [review]
Small issue with the build context. But looks promising!

File: hack/ci/windows.ps1 (hunk 1 above)
Comment (thaJeztah) [reply hack/ci/windows.ps1]: oh, good catch 🤦‍♂

Comment (thaJeztah) [reply hack/ci/windows.ps1]: fixed 👍

File: contrib/busybox/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# To build: docker build -t busybox .
+# To publish: Needs someone with publishing rights
+ARG WINDOWS_BASE_IMAGE=mcr.microsoft.com/windows/servercore
+ARG WINDOWS_BASE_IMAGE_TAG=ltsc2019
+
+FROM ${WINDOWS_BASE_IMAGE}:${WINDOWS_BASE_IMAGE_TAG}
+RUN mkdir C:\tmp && mkdir C:\bin
+ADD http://frippery.org/files/busybox/busybox.exe /bin/
+RUN setx /M PATH "C:\bin;%PATH%"
+RUN powershell busybox.exe --list ^|%{$nul = cmd /c mklink C:\bin\$_.exe busybox.exe}
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply contrib/busybox/Dockerfile]: looks like the current variations assume it's always powershell https://github.com/jhowardmsft/busybox/blob/v1.1/Dockerfile

https://github.com/jhowardmsft/busybox/blob/v1.0/Dockerfile

[the remark this answers is not in the document]
Comment (thaJeztah) [reply contrib/busybox/Dockerfile]: That said; with the Dockerfile in this repo, it'll be easier to modify it if we switch to nanoserver (which we unfortunately no longer can do for the RS1 / Server 2016 CI)

**Comment by thaJeztah** [comment]
Looks like the RS1 node went offline; https://ci.docker.com/public/job/moby/job/PR-39807/4/execution/node/61/log/

```
00:13:10  azwin-2-712180 was marked offline: Connection was broken: java.util.concurrent.TimeoutException: Ping started at 1566943750310 hasn't completed by 1566943990311
00:13:10  	at hudson.remoting.PingThread.ping(PingThread.java:134)
00:13:10  	at hudson.remoting.PingThread.run(PingThread.java:90)
00:13:10  
```

Also the Blue Ocean UI still has a lot of bugs; it's showing the output of the Janky stage when you click on the Windows RS1 stage 😞 

Screenshot 2019-08-28 at 02 06 58

File: contrib/busybox/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (StefanScherer) [reply contrib/busybox/Dockerfile]: We discussed using nanoserver to speed up the CI (faster docker pull, ...), but as nanoserver is only a semi-annual release it's not suitable for 2016/RS1 and is also only short-lived for 2019/RS5 (updates may end next year).

**Comment by ddebroy** [comment]
cc @vikramhh and @ameyag

**Comment by vikramhh** [comment]
LGTM

File: contrib/busybox/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# John Howard (IRC jhowardmsft, Email john.howard@microsoft.com)
+#
+# To build: docker build -t busybox .
+# To publish: Needs someone with publishing rights
+ARG WINDOWS_BASE_IMAGE=mcr.microsoft.com/windows/servercore
+ARG WINDOWS_BASE_IMAGE_TAG=ltsc2019
+
+FROM ${WINDOWS_BASE_IMAGE}:${WINDOWS_BASE_IMAGE_TAG}
+RUN mkdir C:\tmp && mkdir C:\bin
+ADD http://frippery.org/files/busybox/busybox.exe /bin/
```
Comment (justincormack) [inline contrib/busybox/Dockerfile]: Installing some random exe from an http site does not seem  acceptable. `busybox` appears to be available from `choco` can we use that?

Comment (StefanScherer) [reply contrib/busybox/Dockerfile]: This URL is the "official" one to get the Windows port of busybox.
The Chocolatey package `busybox` also uses it https://github.com/chtof/chocolatey-packages/blob/master/automatic/busybox/update.ps1#L13

But, we could use https.

Comment (andrewhsu) [reply contrib/busybox/Dockerfile]: Using https would be better, but if we wanted to guarantee that we're locked to a specific exe that doesn't change under our feet, perhaps grab the specific version, e.g. https://frippery.org/files/busybox/busybox-w32-FRP-3244-g48128b9aa.exe (if you really need to get the 32-bit ver) and `sha256sum` check it to hard-coded `068744fbe3308807c7658da711821af13af6563dab6d86f7aca1ac45e3710eba`

**Comment by thaJeztah** [comment]
ping @tianon @StefanScherer @cpuguy83 PTAL - I think I addressed the review comments (pinned to a fixed version and verifying the sha256)

**Review by tianon — APPROVED** [review]
LGTM

**Review by StefanScherer — APPROVED** [review]
LGTM

**Pull request merged**